pulsar-broker/src/main/java/org/apache/pulsar/broker/lookup/TopicLookupBase.java

```diff
@@ -80,55 +80,63 @@ protected void internalLookupTopicAsync(TopicName topicName, boolean authoritati
-                if (log.isDebugEnabled()) {
-                    log.debug("Lookup succeeded for topic {} -- broker: {}", topicName, result.getLookupData());
+                LookupResult result = optionalResult.get();
+                // We have found either a broker that owns the topic, or a broker to
+                // which we should redirect the client to
+                if (result.isRedirect()) {
+                    boolean newAuthoritative = result.isAuthoritativeRedirect();
+                    URI redirect;
+                    try {
+                        String redirectUrl = isRequestHttps() ? result.getLookupData().getHttpUrlTls()
+                                : result.getLookupData().getHttpUrl();
+                        checkNotNull(redirectUrl, "Redirected cluster's service url is not configured");
+                        String lookupPath = topicName.isV2() ? LOOKUP_PATH_V2 : LOOKUP_PATH_V1;
+                        String path = String.format("%s%s%s?authoritative=%s",
+                                redirectUrl, lookupPath, topicName.getLookupName(), newAuthoritative);
+                        path = listenerName == null ? path : path + "&listenerName=" + listenerName;
+                        redirect = new URI(path);
+                    } catch (URISyntaxException | NullPointerException e) {
+                        log.error("Error in preparing redirect url for {}: {}", topicName, e.getMessage(), e);
+                        completeLookupResponseExceptionally(asyncResponse, e);
+                        return;
+                    }
+                    if (log.isDebugEnabled()) {
+                        log.debug("Redirect lookup for topic {} to {}", topicName, redirect);
+                    }
+                    completeLookupResponseExceptionally(asyncResponse,
+                            new WebApplicationException(Response.temporaryRedirect(redirect).build()));
+
+                } else {
+                    // Found broker owning the topic
+                    if (log.isDebugEnabled()) {
+                        log.debug("Lookup succeeded for topic {} -- broker: {}", topicName, result.getLookupData());
+                    }
+                    completeLookupResponseSuccessfully(asyncResponse, result.getLookupData());
                 }
-                completeLookupResponseSuccessfully(asyncResponse, result.getLookupData());
-            }
-        }).exceptionally(exception -> {
-            log.warn("Failed to lookup broker for topic {}: {}", topicName, exception.getMessage(), exception);
-            completeLookupResponseExceptionally(asyncResponse, exception);
-            return null;
+            }).exceptionally(exception -> {
+                log.warn("Failed to lookup broker for topic {}: {}", topicName, exception.getMessage(), exception);
+                completeLookupResponseExceptionally(asyncResponse, exception);
+                return null;
+            });
         });
```

Comment:
may be we should add exceptionally in there

Reply:
which completableFuture do you mean?
 The `lookupFuture` already has an exceptionally

Reply:
I got it, you mean the `checkTopicExists` may throw an exception.

Reply:
fixed